### ISSUE TITLE
PNG files are non drawn

### ISSUE THREAD

**Issue by ferencIOS** [podofo/podofo#37]
Hello everybody,

what i'm trying to do is draw a png in a signature box.

```c++
            PdfRect annot_rect(0.0, 0.0, 200.0, 100.0);
            PdfPainter painter;

            painter.SetCanvas(*sigXObject);
            painter.Save();
            painter.Restore();
            auto image = document.CreateImage();
            image->Load(filename);
            
            double left = 0.0;
            double top = 0.0;
            double width = image->GetWidth();
            double height = image->GetHeight();

            double scaleX = annot_rect.GetWidth() / width;
            double scaleY = annot_rect.GetHeight() / height;
            double scale = std::min(scaleX, scaleY);
            
            if (scale < 1.0)
            {
                painter.DrawImage(*image, 0.0, 0.0, scale, scale);
            }
            else
            {
                // Center Image
                double x = (annot_rect.GetWidth() - width) / 2.0;
                double y = (annot_rect.GetHeight() - height) / 2.0;
                painter.DrawImage(*image, x, y);
            }

            painter.FinishDrawing();
```

Basically it is the code of the `podofosign` tool.

The same code allows me to use jpg files instead.

Analyzing the old podofo code I found on `PdfImage::SetImageData`:

```c++
    PdfVariant var;
    m_rRect.ToVariant( var );
    this->GetObject()->GetDictionary().AddKey( "BBox", var );

    this->GetObject()->GetStream()->Set( pStream, vecFilters );
```

So I tried adding these new lines on `PdfImage::SetDataRaw`:

```c++
    PdfArray arr;
    GetRect().ToArray(arr);
    auto var = PdfVariant(arr);
    dict.AddKey( "BBox", var );


    PdfFilterList filters;
    filters.push_back( PdfFilterType::FlateDecode );

    if (info.Filters.size() == 0)
        GetObject().GetOrCreateStream().SetData(stream, filters, true);
    else
        GetObject().GetOrCreateStream().SetData(stream, info.Filters, true);
```

the result, however, has not changed.

```c++
#define PNG_LIBPNG_VER_STRING "1.6.39"
#define PNG_HEADER_VERSION_STRING " libpng version 1.6.39 - November 20, 2022\n"
```

Thanks for your help!

Attached:

[blank-signed-jpg.pdf](https://github.com/podofo/podofo/files/10530412/blank-signed-jpg.pdf)
[blank-signed-png.pdf](https://github.com/podofo/podofo/files/10530414/blank-signed-png.pdf)
![image](https://user-images.githubusercontent.com/39296224/215344920-e36d66b6-0e40-471e-82c2-1d5bb75ce544.jpg)
![image](https://user-images.githubusercontent.com/39296224/215344929-283239de-de6b-4ee5-b3a3-bc5fc71fe4b6.png)

**Comment by ceztko** [comment]
> So I tried adding these new lines on PdfImage::SetDataRaw:

0.9.x was doing some things that were of no meaning for the PDF specification, for example a PDF image has not `BBox` entry.

If this is a regression, the best thing is attaching a PDF signed with 0.9.x, or just with a field and the image as in this case.

**Comment by ferencIOS** [comment]
It might be the color space:

**podofo-next**
```c++
14 1 obj
<</Type/XObject/BitsPerComponent 8/ColorSpace[/DeviceRGB 22 15 1 R]/Filter/FlateDecode/Height 1264/Length 370865/Subtype/Image/Width 1353>>
stream
[...]
endstream
endobj
```

**0.9.x**
```c++
35 0 obj<</Type/XObject/BBox[ 0 0 1174 1557]/BitsPerComponent 8/ColorSpace[/Indexed/DeviceRGB 22 36 0 R]/Filter/FlateDecode/Height 1557/Length 401854/Subtype/Image/Width 1174>>
stream
[...]
endstream
endobj
```

So i took the [documentation](https://opensource.adobe.com/dc-acrobat-sdk-docs/pdfstandards/PDF32000_2008.pdf
):

```c++
An Indexed colour space shall be defined by a four-element array:
[/Indexed base hival lookup]

 /Indexed
/DeviceRGB
255
<000000 FF0000 00FF00 0000FF B57342 …>
]
```

And in the code I found an interesting thing:

```c++
        PdfArray arr;
        arr.Add(PdfName(PoDoFo::ColorSpaceToNameRaw(info.ColorSpace)));
        arr.insert(arr.begin(), info.ColorSpaceArray.begin(), info.ColorSpaceArray.end());
        dict.AddKey("ColorSpace", info.ColorSpaceArray);
```

and replacing with
```c++
        PdfArray arr(info.ColorSpaceArray);
        arr.insert(arr.begin(), PdfName(PoDoFo::ColorSpaceToNameRaw(info.ColorSpace)));
        dict.AddKey("ColorSpace", arr);
```

it works!

[blank-signed.pdf](https://github.com/podofo/podofo/files/10531208/blank-signed.pdf)
[blank-signed-IT-WORKS.pdf](https://github.com/podofo/podofo/files/10531320/blank-signed-IT-WORKS.pdf)

**Comment by ceztko** [comment]
Ooops, I needed one more commit ;) https://github.com/podofo/podofo/commit/994990dee6cc624266560461adbfbf40e4a0f79b

**Comment by ceztko** [comment]
Thank you for the fix! I reworked a lot `PdfImage` data setting to be cleaner and consistent, for example one now can can re-set the data in different ways on the same image and you won't/shouldn't end in inconsistent settings being applied, which was very possible in PoDoFo 0.9.x. Unfortunately I didn't have example images to test such code paths. Also unfortunately the whole `/ColorSpace` situation is confusing and difficult to unit test. Let say for example that these custom color space definitions make sense, as it was in the code we found from 0.9.x, and correctly produce different colors when rendered to screen with eg. Adobe Reader: to unit test this I should be able to process the whole color pipeline and apply a transformation to the raw pixels as extracted from the PNG and have the true render colors. This is not done as we don't have this color pipeline at all. Possibly it's a render thing and would be best suited in render sharers, and potentially slow to implement in regular CPU code. Doing a color pipeline processing would require some relevant expertise.
